### ISSUE TITLE
Batch Normalization layers inconsistently placed in regards to paper

### ISSUE THREAD

**Issue by zzuberi** [srebuffi/residual_adapters#2]
According to "Efficient parametrization of multi-domain deep neural networks" the Batch Normalization (BN) layers are not needed for the parallel configuration but the model it appears that you still use task specific BN layers for the parallel configuration:

`elif config_task.mode == 'parallel_adapters' and is_proj:
            self.parallel_conv = nn.ModuleList([conv1x1(in_planes, planes, stride) for i in range(nb_tasks)])
            self.bns = nn.ModuleList([nn.BatchNorm2d(planes) for i in range(nb_tasks)])`

Additionally in the series configurations, the BN layer is stated to come before the convolution yet that is not the case in the provided code:

`if config_task.mode == 'series_adapters' and is_proj:
            self.bns = nn.ModuleList([nn.Sequential(conv1x1(planes), nn.BatchNorm2d(planes)) for i in range(nb_tasks)])`

**Comment by liuquande** [comment]
I also have the concern about the BN configurations inconsistency with paper.

**Comment by emergencyd** [comment]
is this question answered yet? I am also curious about it

**Comment by srebuffi** [comment]
I thank you for your comment. Indeed, the paper could have been a bit clearer about this point. 

First, for the parallel configuration, specific BN layers are needed as the data distribution is very different from one dataset to another (hence mean and std cannot be shared). We should have put the BN layer in blue in the Figure 2,b of the paper to make it clearer.

Second, in the series configuration, there are two BN layers as `conv1x1(planes) = Identity + nn.Sequential(nn.BatchNorm2d(planes), conv1x1_fonc(planes))`. Though, I advice against the series configuration as it requires pretraining with the adapters already in the network whereas the parallel configuration is plug and play with networks pretrained and available online.

**Comment by emergencyd** [comment]
Thank you for your reply! but why cannot the series adapter be plug and play?

**Comment by srebuffi** [comment]
We tried the series configuration in a plug and play configuration but the model did not train properly. I assume that inserting modules in series in a neural network hurts the backpropagation process whereas inserting these same modules seem to preserve good training properties. I do not have a clear reason why it works that way, it is just an observation.

**Comment by emergencyd** [comment]
Quite clear now! Thank you so much!
